OrbitCore/LinuxEventTracer.cpp

```diff
@@ -6,89 +6,102 @@
+#include "OrbitProcess.h"
+#include "OrbitModule.h"
 #include "Params.h"
+#include "PrintVar.h"
+#include "SamplingProfiler.h"
 #include "TimerManager.h"
-#include "ContextSwitch.h"
-#include "CoreApp.h"
-
-#include <vector>
-
 
 //-----------------------------------------------------------------------------
 void LinuxEventTracer::Start()
 {
     PRINT_FUNC;
 #if __linux__
-    m_ExitRequested = false;
+    *m_ExitRequested = false;
 
-    m_Thread = std::make_shared<std::thread>(&LinuxEventTracer::Run, &m_ExitRequested);
+    m_Thread = std::make_shared<std::thread>(&LinuxEventTracer::Run,
+            m_Pid, m_Frequency, m_ExitRequested);
     m_Thread->detach();
-
 #endif
 }
 
 //-----------------------------------------------------------------------------
 void LinuxEventTracer::Stop()
 {
-    m_ExitRequested = true;
+    *m_ExitRequested = true;
 }
 
-void LinuxEventTracer::Run(bool* a_ExitRequested)
+void LinuxEventTracer::Run(pid_t a_Pid, int32_t a_Frequency,
+    const std::shared_ptr<std::atomic<bool>>& a_ExitRequested)
 {
-    // get the number of cpus
-    int cpus = std::thread::hardware_concurrency();
-    // some compilers does not support hardware_concurrency.
-    if (cpus == 0)
+    int cpus = static_cast<int>(std::thread::hardware_concurrency());
+    // Some compilers does not support hardware_concurrency.
+    if (cpus == 0) {
         cpus = std::stoi(LinuxUtils::ExecuteCommand("nproc"));
+    }
     
-    std::vector<uint32_t> fds;
+    std::vector<int32_t> fds;
 
-    if ( GParams.m_SystemWideScheduling )
+    if (GParams.m_TrackContextSwitches)
     {
-        // open perf events for all cpus to keep track of process spawning    
-        for (int cpu = 0; cpu < cpus; cpu++)
+        if ( GParams.m_SystemWideScheduling )
+        {
+            // perf_event_open for all cpus to keep track of process spawning.
+            for (int cpu = 0; cpu < cpus; cpu++)
+            {
+              fds.push_back(LinuxPerfUtils::context_switch_open(-1, cpu));
+            }
+        }
+        else
         {
-            fds.push_back(LinuxPerfUtils::context_switch_open(-1, cpu));
+            // perf_event_open for all cpus and the PID to keep track of process spawning.
+            fds.push_back(LinuxPerfUtils::context_switch_open(Capture::GTargetProcess->GetID(), -1));
         }
     }
-    else 
-    {
-        // open_perf_events for all cpus and the PID to keep track of process spawning
-        fds.push_back(LinuxPerfUtils::context_switch_open(Capture::GTargetProcess->GetID(), -1));
+
+    for (pid_t tid : LinuxUtils::ListThreads(Capture::GTargetProcess->GetID())) {
+        // Keep threads in sync.
+        Capture::GTargetProcess->AddThreadId(tid);
+
+        if (!GParams.m_SampleWithPerf) {
+            fds.push_back(LinuxPerfUtils::stack_sample_event_open(
+                    tid, a_Frequency));
+        }
     }
 
+    LibunwindstackUnwinder unwinder{};
+    unwinder.SetMaps(LinuxUtils::ReadMaps(a_Pid));
```

Comment:
Check error case here?

Reply:
For now I'm not sure, as an error should only happen if the process does not exist anymore, but recovering from such error is beyond the scope of this change.